### PR TITLE
Forward timestamps

### PR DESCRIPTION
**Pull request by bartkrak**
partially solves: https://github.com/membraneframework/membrane_core/issues/735

### CODE REVIEW

**Comment by spscream** [comment]
I have crash with the following pipeline opus_decode -> convert 48khz to 32khz -> g7221 encoder -> rtp, this merge fixes it. Thanks

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex [hunk 1]
```diff
@@ -122,7 +122,13 @@ defmodule Membrane.FFmpeg.SWResample.Converter do
   end
 
   @impl true
-  def handle_buffer(:input, %Buffer{payload: payload}, _ctx, state) do
+  def handle_buffer(:input, %Buffer{payload: payload, pts: pts}, _ctx, state) do
+    state =
+      state
+      |> Map.merge(%{
+        current_pts: pts
+      })
```
Comment (FelonEkonom) [inline lib/membrane_ffmpeg_swresample_plugin/converter.ex:130]:
```suggestion
|> Map.put(:current_pts, pts)
```

**Review by FelonEkonom — CHANGES_REQUESTED** [review]
Not sure, if passing `pts` from input to output is a good approach here, I recommend to ask somebody more familiar with this plugin, if it doesn't require more sophisticated logic (e.g. passing on output `pts` of the first buffer, which payload was part of `state.queue`)

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex (hunk 1 above)
Comment (mat-hek) [reply lib/membrane_ffmpeg_swresample_plugin/converter.ex]: @bartkrak actually, the right approach would be to add this field in `handle_init` and do `%{state | current_pts: pts}`. Adding fields to state in various callbacks makes it hard to follow.

**Review by mat-hek — CHANGES_REQUESTED** [review]
As we spoke, we must take into account the fact that the resampler may queue the stream

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex [hunk 2]
```diff
@@ -122,20 +123,29 @@ defmodule Membrane.FFmpeg.SWResample.Converter do
   end
 
   @impl true
-  def handle_buffer(:input, %Buffer{payload: payload, pts: pts}, _ctx, state) do
+  def handle_buffer(:input, %Buffer{payload: payload, pts: input_pts}, _ctx, state) do
+    input_frame_size = RawAudio.frame_size(state.input_stream_format)
+    output_frame_size = RawAudio.frame_size(state.output_stream_format)
+    ratio = output_frame_size / input_frame_size
+
+    expected_output_frames_count = byte_size(payload) / input_frame_size * ratio
```
Comment (mat-hek) [inline lib/membrane_ffmpeg_swresample_plugin/converter.ex:131]: I think we should round to an integer here

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex [hunk 3]
```diff
@@ -218,4 +233,21 @@ defmodule Membrane.FFmpeg.SWResample.Converter do
...
+
+  defp update_pts_queue(state, converted_frames_count) do
+    {[{out_pts, expected_frames}], rest} = Enum.split(state.pts_queue, 1)
+
+    cond do
+      converted_frames_count < expected_frames ->
+        state =
+          Map.update!(state, :pts_queue, fn _pts_queue ->
+            [{out_pts, expected_frames - converted_frames_count}] ++ rest
+          end)
```
Comment (mat-hek) [inline lib/membrane_ffmpeg_swresample_plugin/converter.ex:245]: this can be `%{state | pts_queue: ...`

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex [hunk 4]
```diff
@@ -218,4 +233,21 @@ defmodule Membrane.FFmpeg.SWResample.Converter do
...
+
+        {state, out_pts}
+
+      converted_frames_count >= expected_frames ->
+        {%{state | pts_queue: rest}, out_pts}
+    end
```
Comment (mat-hek) [inline lib/membrane_ffmpeg_swresample_plugin/converter.ex:251]: this seems like an if

File: test/membrane_ffmpeg_swresample_plugin/converter_test.exs [hunk 5]
```diff
@@ -157,7 +157,14 @@ defmodule Membrane.FFmpeg.SWResample.ConverterTest do
 
       assert {[], new_state} = @module.handle_buffer(:input, buffer, nil, state)
 
-      assert new_state == %{state | queue: payload}
+      # this is necesary to ignore state.pts_queue in the assertion
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/converter_test.exs:160]: you can set it to `nil` on both sides then

File: test/membrane_ffmpeg_swresample_plugin/converter_test.exs [hunk 6]
```diff
@@ -178,5 +186,21 @@ defmodule Membrane.FFmpeg.SWResample.ConverterTest do
       assert actions == [buffer: {:output, %Membrane.Buffer{payload: result}}]
       assert new_state == %{state | queue: <<0::2*8>>}
     end
+
+    # test "timestamps forward test", %{state: initial_state} do
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/converter_test.exs:190]: leftover

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  import Membrane.ChildrenSpec
+  import Membrane.Testing.Assertions
+
+  alias Membrane.FFmpeg.SWResample.Converter
+  alias Membrane.{RawAudio, Testing}
+
+  test "pts forward test" do
+    input_stream_format = %RawAudio{sample_format: :s16le, sample_rate: 16_000, channels: 2}
+    # input_stream_format = %RawAudio{sample_format: :u8, sample_rate: 8_000, channels: 1}
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:12]: more leftovers

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex [hunk 8]
```diff
@@ -128,7 +128,7 @@ defmodule Membrane.FFmpeg.SWResample.Converter do
     output_frame_size = RawAudio.frame_size(state.output_stream_format)
     ratio = output_frame_size / input_frame_size
 
-    expected_output_frames_count = byte_size(payload) / input_frame_size * ratio
+    expected_output_frames_count = (byte_size(payload) / input_frame_size * ratio) |> trunc()
```
Comment (mat-hek) [inline lib/membrane_ffmpeg_swresample_plugin/converter.ex:131]: I think it's necessary to use `round`. Otherwise we'd expect 2137 frames if the float result is 2137.99

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+  alias Membrane.FFmpeg.SWResample.Converter
+  alias Membrane.{RawAudio, Testing}
+
+  test "pts forward test" do
+    input_stream_format = %RawAudio{sample_format: :s16le, sample_rate: 16_000, channels: 2}
+    output_stream_format = %RawAudio{sample_format: :s32le, sample_rate: 32_000, channels: 2}
+
+    # 32 frames * 2048 bytes
+    fixture_path = Path.expand(Path.join(__DIR__, "/../fixtures/input_s16le_stereo_16khz.raw"))
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:15]: the run path in tests is always the project root, so it's safe to do
```suggestion
    fixture_path = "test/fixtures/input_s16le_stereo_16khz.raw"
```

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      |> child(:sink, Membrane.Testing.Sink)
+    ]
+
+    pipeline = Testing.Pipeline.start_link_supervised!(spec: spec)
+    assert_start_of_stream(pipeline, :sink)
+    assert_sink_buffer(pipeline, :sink, _buffer)
+
+    Enum.each(0..30, fn index ->
+      assert_sink_buffer(pipeline, :sink, %Membrane.Buffer{pts: out_pts})
+      assert out_pts == index * 31_250_000
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:32]: let's make this magic number a variable

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    end)
+
+    assert_sink_buffer(pipeline, :sink, %Membrane.Buffer{pts: 937_500_000})
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:35]: why separate assertion for this one?

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        payload: payload,
+        pts: index * 31_250_000
+      }
+    end)
+  end
+
+  @spec split_binary(binary(), list(binary())) :: list(binary())
+  def split_binary(binary, acc \\ [])
+
+  def split_binary(<<binary::binary-size(2048), rest::binary>>, acc) do
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:57]: Would it be hard to have the chunk length differ from buffer to buffer? I guess we could have better coverage that way

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs (hunk 11 above)
Comment (bartkrak) [reply test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:35]: Because of the way converter sets pts. In this exact case, last two buffers have the same pts, so there can be a separate assert outside of the loop, or I would have to make a condition inside the loop above that checks if It's on the last index, to assert with pts the same as previous one. Don't know which solution is prettier 🧐

Comment (mat-hek) [reply test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:35]: Ok. Why do they have the same PTS?

Comment (bartkrak) [reply test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:35]: because converter buffers some data while resampling, last two buffers are a bit smaller in size, and both are made from data that came with the same pts. (algorithm that we talked about on slack)

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs (hunk 12 above)
Comment (mat-hek) [reply test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:57]: but if it's too much work let's leave it as is

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+  defp buffers_from_file(path) do
+    binary = File.read!(path)
+
+    split_binary(binary)
+    |> Enum.with_index()
+    |> Enum.map(fn {payload, index} ->
+      %Membrane.Buffer{
+        payload: payload,
+        pts: index * 31_250_000
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:51]: actually pts_multiplayer can be a module attribute and you can use it here as well

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    assert_sink_buffer(pipeline, :sink, _buffer)
+
+    pts_multiplier = 31_250_000
+
+    Enum.each(0..30, fn index ->
+      assert_sink_buffer(pipeline, :sink, %Membrane.Buffer{pts: out_pts})
+      assert out_pts == index * pts_multiplier
+    end)
+
+    assert_sink_buffer(pipeline, :sink, %Membrane.Buffer{pts: 937_500_000})
```
Comment (mat-hek) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:37]:
```suggestion
    assert_sink_buffer(pipeline, :sink, %Membrane.Buffer{pts: 30 * pts_multiplier)
```

**Review by mat-hek — APPROVED** [review]

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs (hunk 14 above)
Comment (bartkrak) [reply test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:37]: using multiplication inside a match throws compile error

File: test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+      |> child(:resampler, %Converter{
+        input_stream_format: input_stream_format,
+        output_stream_format: output_stream_format
+      })
+      |> child(:sink, Membrane.Testing.Sink)
+    ]
+
+    pipeline = Testing.Pipeline.start_link_supervised!(spec: spec)
+    assert_start_of_stream(pipeline, :sink)
+    assert_sink_buffer(pipeline, :sink, _buffer)
```
Comment (FelonEkonom) [inline test/membrane_ffmpeg_swresample_plugin/pts_forward_test.exs:30]: The second assertion makes the first one redundant

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex [hunk 16]
```diff
@@ -122,16 +123,29 @@ defmodule Membrane.FFmpeg.SWResample.Converter do
   end
 
   @impl true
-  def handle_buffer(:input, %Buffer{payload: payload}, _ctx, state) do
+  def handle_buffer(:input, %Buffer{payload: payload, pts: input_pts}, _ctx, state) do
+    input_frame_size = RawAudio.frame_size(state.input_stream_format)
+    output_frame_size = RawAudio.frame_size(state.output_stream_format)
+    ratio = output_frame_size / input_frame_size
+
+    expected_output_frames_count = (byte_size(payload) / input_frame_size * ratio) |> round()
```
Comment (FelonEkonom) [inline lib/membrane_ffmpeg_swresample_plugin/converter.ex:131]: It is more precise to write `(byte_size(payload) * output_frame_size / (input_frame_size * input_frame_size))` because of limitted accuracy of floats

File: lib/membrane_ffmpeg_swresample_plugin/converter.ex [hunk 17]
```diff
@@ -212,4 +233,15 @@ defmodule Membrane.FFmpeg.SWResample.Converter do
       {:error, reason} -> raise "Error while flushing converter: #{inspect(reason)}"
     end
   end
+
+  defp update_pts_queue(state, converted_frames_count) do
+    {[{out_pts, expected_frames}], rest} = Enum.split(state.pts_queue, 1)
```
Comment (FelonEkonom) [inline lib/membrane_ffmpeg_swresample_plugin/converter.ex:238]: You can just match
```suggestion
    [{out_pts, expected_frames} | rest] = state.pts_queue
```

**Review by FelonEkonom — CHANGES_REQUESTED** [review]

**Review by FelonEkonom — APPROVED** [review]

**Review by FelonEkonom — APPROVED** [review]
